external/dir.props: remove Portable condition

tmds opened the pull request:
This PR checks what platforms fail when removing the PortableBuild condition in external/dir.props

ericstj commented:
You're going to provide a much more detailed explanation about why you think you need to do this.

Explain what problem you are trying to solve, how this solves it, and what alternatives you considered.

We've spent a lot of time trying to simplify the number of RIDs floating around and how they are calculated, which take precedence, etc.  This appears to be a step backwards.

ericstj commented on config.json:
```diff
@@ -31,11 +31,29 @@
       "defaultValue": false
     },
     "RuntimeOS": {
```
This is meant to represent what we're running on.

ericstj commented on config.json:
```diff
@@ -31,11 +31,29 @@
+      "values": ["win7", "osx.10.10", "ubuntu.14.04", "ubuntu.16.04", "etc-other-rid-based-os-names"],
+      "defaultValue": ""
+    },
+    "TargetRid": {
```
This already exists today and is PackageRID.

tmds commented:
@ericstj sure. Sorry for the lack of context. This is part of https://github.com/dotnet/corefx/issues/20731
The use case is: bootstrapping a build by a package maintainer.
I started this PR to get a better understanding of how things fit together. I'm sure things can be improved a lot.
The PR as it is currently splits the RuntimeOS in 3 properties:
* HostRID: can be set to override the rid of the build system, e.g. tizen build needs this since there is no tizen x64 ilasm
* RuntimeRid: used to select the runtime
* TargetRid: this allows to name the build output

ericstj commented:
Why couldn't you use the existing properties?

> HostRID: can be set to override the rid of the build system, e.g. tizen build needs this since there is no tizen x64 ilasm

Don't plumb a new global property to deal with this.  It looks to me like you're trying to build for a new RID: tizen-x64.  The problem is that RID is missing from https://github.com/dotnet/corefx/blob/master/pkg/Microsoft.NETCore.Platforms/runtime.json#L366-L374.  If it were present, then tizen would fallback to linux-x64 and you'd get that by default (after the package was published, and consumed by these projects).

> RuntimeRid: used to select the runtime

This already exists, it's `RuntimeOS`.  Architecture is passed separately.  Why doesn't that work for you?

>  TargetRid: this allows to name the build output

This already exists, it's `PackageRID`.

Thinking about the whole scenario (building on a new RID), suppose you just specified RuntimeOS to some existing RID that was supported in order to bootstrap off of some existing binaries for toolchain, and specified PackageRID as your new RID?

tmds commented:
@ericstj 
I'm part of the team that maintains the .NET Core package for RHEL and part of the .NET SIG that maintains the package for Fedora.

So far, my experience regarding building .NET Core and RIDs is that:
- all dotnet repos contain a case just below reading /etc/os-release that says: if this is rhel.7.x, then strip the minor.
- corefx runtime.json contains fedora.26 to be sure we may be capable of building .NET Core on the next version of Fedora.

When bootstrapping:

If you try and do a non-portable build on an unsupported (i.e. not part of Tools' runtime.json) machine, it will fail.
First it will fail to pull external/ilasm, which can be fixed by RuntimeOS. And then it will fail to fetch external/runtime
because it is using the ilasm RID to fetch the runtime. This is what lead to splitting into HostRid and RuntimeRid.

To do a decent build, one would actually want to do two builds. The first build would be based on a 3rd party cli (e.g. Microsoft linux-x64 build).
This would generate the stage1. Then a second build would use the stage1 build the final build. This ensures that the package isn't built using 3rd party binaries.
This lead me to the TargetRid property.
e.g. the two stages on a linux-x64 compatible unknown banana.28-x64 linux would be:

```
build coreclr firstbuild
corefx: HostRID=linux-x64 RuntimeRID=firstbuild TargetRID=firstbuild

build coreclr banana.28 using firstbuild
corefx: HostRID=firstbuild RuntimeRID=banana.28 TargetRID=banana.28-x64
```

(Other than that, the TargetRid is usefull to handle the cases like the rhel renaming.)

Not included in this PR is a build task which updates runtime.json with the TargetRID if it is not yet present.
There is such a task as part of the source build project.

I should have shared this before, but I didn't expect you to be added as a code reviewer. I was using this PR to improve my understanding and experiment using the CI setup (since I only build for Linux locally).

I hope this properly explains the use-case. If you want, I can close the PR and update the issue (https://github.com/dotnet/corefx/issues/20731) with this text or we can continue the discussion here.

ericstj commented:
/cc @ellismg and @chcosta who are working on the vertical build scenario.

> First it will fail to pull external/ilasm, which can be fixed by RuntimeOS. And then it will fail to fetch external/runtime
because it is using the ilasm RID to fetch the runtime. This is what lead to splitting into HostRid and RuntimeRid.

You should be able to use a single RuntimeOS to select both a working ILASM and runtime.  Please be **specific** in describing a case where this doesn't work. If that is so, fix the bug causing that case.  Don't plumb more workarounds into the infrastructure.

As far as I can tell TargetRID seems functionally equivalent to PackageRID.  Let me know if I am missing something here. Please be **specific** in describing why this doesn't work.  Likely its better to just fix bugs there than add new concepts...

It also sounds like you're trying to encode an intermediate phase into the runtimeID.  Please don't do that.  You can just as well manage that through paths/versioning which is consistent with how we normally flow bits across the repositories.

tmds commented:
I made this PR to improve my understanding of the code and experiment. Some of these things I am trying will be possible by using (or extending) existing properties. This wasn't meant as a final PR.

> You should be able to use a single RuntimeOS to select both a working ILASM and runtime. Please be specific in describing a case where this doesn't work. If that is so, fix the bug causing that case. Don't plumb more workarounds into the infrastructure.

This is when the rid of the host os is not in the Tools runtime.json. This is the case for every 'new' OS. It's similar to why Tizen has this line in ilasm.depproj: https://github.com/dotnet/corefx/blob/master/external/ilasm/ilasm.depproj#L8. 

> As far as I can tell TargetRID seems functionally equivalent to PackageRID. Let me know if I am missing something here. Please be specific in describing why using PackageRID doesn't work. Likely its better to just fix bugs there than add new concepts...

This may be a perfect match.

> It also sounds like you're trying to encode an intermediate phase into the runtimeID. Please don't do that. You can just as well manage that through paths/versioning which is consistent with how we normally flow bits across the repositories.

Agree.

ericstj commented:
> I made this PR to improve my understanding of the code and experiment. Some of these things I am trying will be possible by using (or extending) existing properties. This wasn't meant as a final PR.

Understood.

The Tizen case seems like a mistake.  Looking at https://github.com/dotnet/corefx/pull/18725 it looks like @jyoungyun was specifying RuntimeOS when TargetRID would have been more appropriate.  This is exactly why I don't want this stuff to get more complicated, its already complicated enough 😃 

I would expect bootstrapping to look like:
build -- /p:RuntimeOS=<someExistingOS>;TargetRID=<newRID>

For example <someExistingOS> here would be "linux" or "ubuntu" in order to use all the runtime assets from that existing OS.  Suppose you were trying to use some brand new RID here that you just built in CoreClr, it'd be a bug in ILAsm package if you built for that RID and the built package out of CoreCLR didn't support it.

TargetRID is the new RID you're building for.

jyoungyun commented:
@ericstj @tmds The reason why I added https://github.com/dotnet/corefx/pull/21032 is that Tizen has not published x64 packages yet according to our policy. So, I use ubuntu packages as fallback source. If we already have x64 packages, we do not need to add that line.

So, I think in Tizen case, it seems more appropriate to use `RuntimeOS` instead of `targetRID`.

If you have any other comments, please tell me. :)

ericstj commented:
@jyoungyun, if you use `RuntimeOS=ubuntu.14.04;TargetRID=tizen.4.0.0-armel` wouldn't that work without a special case in the ILAsm project?

tmds commented:
@ericstj I've adjusted the PR based on the discussion.

* It is adding one new config option `HostOS`. This can be used when the cli under Tools does not know RuntimeOS (e.g. brand new linux flavor). On your open question to @jyoungyun on this: I think `RuntimeOS=ubuntu.14.04;TargetRID=tizen.4.0.0-armel` does not work for because that would restore `ubuntu.14.04-armel` in `runtime.depproj` where this must be `tizen.4.0.0-armel`.
* The PR also refactors the use of RIDs: it centralizes the logic in `dir.props` and refers to `HostRid` and `TargetRid` from other places.
* In `dir.props` I tried to calculate the portable RID based on `OSGroup` but this fails in CI, I don't understand why (see commended out XML).

ericstj commented on dir.props:
```diff
@@ -117,6 +117,28 @@
+    <_portableRID Condition="'$(_runtimeOSFamily)' == 'win'">win</_portableRID>
+    <_portableRID Condition="'$(_runtimeOSFamily)' == 'osx'">osx</_portableRID>
+<!--
+    Why doesn't this work?
```
OSGroup is specific to the project configuration.  For most projects which don't cross-compile by OSGroup this will be "AnyOS".

ericstj replied:
This is the "AllConfigurations" build concept.  We need to maintain that build, but it doesn't need to support TargetRID.  TargetRID concept clashes with "AllConfigurations" concept.

ericstj commented on dir.props:
```diff
@@ -117,6 +117,28 @@
+    <HostRid>$(HostOS)-x64</HostRid>
+
+    <_runtimeOSVersionIndex>$(RuntimeOS.IndexOfAny(".-0123456789"))</_runtimeOSVersionIndex>
+    <_runtimeOSFamily Condition="'$(_runtimeOSVersionIndex)' != '-1'">$(RuntimeOS.SubString(0, $(_runtimeOSVersionIndex)))</_runtimeOSFamily>
```
We should make this RuntimeOSFamily if we intend to use it from multiple targets/project files.

ericstj commented on config.json:
```diff
@@ -36,6 +36,12 @@
       "defaultValue": ""
     },
+    "HostOS": {
+      "description": "The RID used for restoring native OS dependent assets (i.e. RID specific binaries)",
```
It's still not clear to me from these two descriptions how this differs from RuntimeOS, we need to make that crisp.

ericstj commented:
> I think RuntimeOS=ubuntu.14.04;TargetRID=tizen.4.0.0-armel does not work for because that would restore ubuntu.14.04-armel in runtime.depproj where this must be tizen.4.0.0-armel.

Then to fix that we should make runtime.depproj honor the TargetRid.

Bottomline I still think we only should have two concepts: what we're *building on* and what we're *building for*, today that's RuntimeOS and TargetRID respectively.  I think each project should decide if it makes sense to use one or the other and behave cleanly in that way.

jyoungyun commented:
When I built by the following command, it failed.

```
docker run -it --rm -v $(pwd):/opt/code -w /opt/code -e 'ROOTFS_DIR=/crossrootfs/armel.tizen.build' hqueue/dotnetcore:ubuntu1404_cross_prereqs_v4-tizen_rootfs ./build.sh -buildArch=armel -release -RuntimeOS=ubuntu.14.04 -PortableBuild=false -- /p:OverridePackageSource=https://tizen.myget.org/F/dotnet-core/api/v3/index.json /p:TargetRID=tizen.4.0.0-armel
```

```
/opt/code/Tools/depProj.targets(83,5): error : Error no assets were resolved from NuGet packages. [/opt/code/external/runtime/runtime.depproj]
    0 Warning(s)
    1 Error(s)
```

Do you mean the RuntimeOS is *what we're building on* and the PackageRID(targetRID) is *what we're building for* ? Then, config.json should support to 'PackageRID' explicitly and runtime.depproj and ilasm.depproj should restore by 'PackageRID'. Please tell me if I need to check before merging it.

ericstj commented:
@jyoungyun there are still issues with the change that @tmds needs to resolve.  I think as part of this change we should prescribe how to build for Tizen on an ubuntu/x64 machine so your test-case here is good validation.

> Do you mean the RuntimeOS is what we're building on and the PackageRID(targetRID) is what we're building for ? 

Yes

> runtime.depproj and ilasm.depproj should restore by 'PackageRID'. 

Partially correct.
`runtime.depproj` should *consider* a PackageRID override if specified since `runtime.depproj` isn't used for build-time execution, but for references from runtime facades and for test execution.
`ilasm.depproj` should restore by `$(RuntimeOS)-<arch>` (where `<arch>` here is the architecture of the build machine, I haven't looked into what the right property is, I think folks have been hardcoding x64) since we need to run ilasm during the build.

ericstj commented on src/SharedFrameworkValidation/RestoreSDKProject/RestoreSDKProject.csproj:
```diff
@@ -4,7 +4,7 @@
     <RuntimeFrameworkVersion>$(_RuntimeFrameworkVersion)</RuntimeFrameworkVersion>
     <CoreFxPackageVersion>4.4.0-$(CoreFxExpectedPrerelease)</CoreFxPackageVersion>
-    <RuntimeIdentifier>$(_PublishRID)</RuntimeIdentifier>
+    <RuntimeIdentifier>$(TargetRid)</RuntimeIdentifier>
```
Looks to me like this was previously using the build-on-RID.  @joperezr can confirm if it's OK to change this to the buiding-for-RID, and double check that you aren't missing anything by removing the global property passing that was here before.

ericstj commented on pkg/Microsoft.Private.CoreFx.UAP/uap.rids.props:
```diff
@@ -1,22 +1,8 @@
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
-    <_aotSuffix Condition="'$(TargetGroup)' == 'uapaot'">-aot</_aotSuffix>
+    <PackageRID>$(TargetRid)</PackageRID>
```
I still see PackageRID used in multiple places where it is just set directly to TargetRID.  A few points: why do the rename? are you sure you every place that needs this remapping? why not just clean up PackageRID entirely?

ericstj commented on dir.props:
```diff
@@ -117,6 +117,26 @@
+    <_portableRID Condition="'$(OSGroup)' == 'Windows_NT'">win</_portableRID>
+    <_portableRID Condition="'$(OSGroup)' == 'OSX'">osx</_portableRID>
+
+    <TargetRid Condition="'$(PortableBuild)' == 'true'">$(_portableRID)-$(ArchGroup)</TargetRid>
```
Typically for these sorts of things we only set them if they've not been set.

joperezr commented on src/SharedFrameworkValidation/SharedFrameworkValidation.proj:
```diff
@@ -57,16 +56,15 @@
 
   <Target Name="BuildPackagesProject">
     <PropertyGroup>
-      <_RestorePackagesCommand>"$(DotnetToolCommand)" restore $(MSBuildThisFileDirectory)/RestoreSDKProject/RestoreSDKProject.csproj /p:_PublishRID=$(_PublishRID) /p:_RuntimeFrameworkVersion=$(MicrosoftNETCoreAppPackageVersion) /p:CoreFxExpectedPrerelease=$(CoreFxExpectedPrerelease)</_RestorePackagesCommand>
```
We still need to pass this in since the project we are building doesn't import dir.props and doesn't know about the regular properties like `TargetRID`

joperezr commented on src/SharedFrameworkValidation/SharedFrameworkValidation.proj:
```diff
@@ -57,16 +56,15 @@
 
   <Target Name="BuildDependenciesProject">
     <MSBuild Projects="../../external/SharedFrameworkValidation/harvestPackages.SharedFramework.depproj"
-             Properties="_PublishRID=$(_PublishRID)"
```
We need to re-add this back because of my previous comment.

joperezr commented on src/SharedFrameworkValidation/SharedFrameworkValidation.proj:
```diff
@@ -10,7 +10,6 @@
     <LKGSharedFrameworkExtractPath>$(RefRootPath)sharedFrameworkZip.LKG/shared/Microsoft.NETCore.App/$(_LKGSharedFrameworkVersion)</LKGSharedFrameworkExtractPath>
     <_RunTestsScriptName Condition="'$(RunningOnUnix)' != 'true'">CloneAndRunTests.cmd</_RunTestsScriptName>
     <_RunTestsScriptName Condition="'$(RunningOnUnix)' == 'true'">CloneAndRunTests.sh</_RunTestsScriptName>
-    <_PublishRID>$(RuntimeOS)-$(ArchGroup)</_PublishRID>
```
All I would change is the definition of _PublishRID here to use TargetRID instead, but all of the other changes on this file and the RestoreSDKProject should be undone.

tmds commented:
@ericstj @joperezr thanks for your feedback, it helps me also to better understand how the pieces fit together.

Current state of PR:
* Added config option to handle case where build system not supported. RuntimeOS represents build system, PackageRID represents target rid.
* Moved RID calc logic into top-level dir.props. All other places either use RuntimeRID (derived from RuntimeOS) or PackageRID. This makes it clear if the RID represents the build / target.

The NETFX build is failing. My guess is this is due to the PackageRID always being set. So probably there should be a condition added so it is not always set. Do you have some input?

joperezr commented on src/SharedFrameworkValidation/SharedFrameworkValidation.proj:
```diff
@@ -57,16 +56,16 @@
   <Target Name="BuildDependenciesProject">
     <MSBuild Projects="../../external/SharedFrameworkValidation/harvestPackages.SharedFramework.depproj"
-             Properties="_PublishRID=$(_PublishRID)"
+            Properties="_PackageRID=$(PackageRID)"
```
Not 100% convinced that the property rename is required, but I'm ok with the changes.

tmds replied:
Thanks. The rename is to make more clear what RIDs are for the build system and which are for the package.

tmds commented:
> The NETFX build is failing. My guess is this is due to the PackageRID always being set. So probably there should be a condition added so it is not always set. Do you have some input?

@ericstj @weshaggard ?

ericstj commented:
This looks busted:
```
  Microsoft.NETCore.Targets -> D:\j\workspace\windows-TGrou---2a8f9c29\bin/packages/Release/specs/Microsoft.NETCore.Targets.nuspec
  Microsoft.NETCore.Platforms -> D:\j\workspace\windows-TGrou---2a8f9c29\bin/packages/Release/specs/Microsoft.NETCore.Platforms.nuspec
  Microsoft.NETCore.Targets -> D:\j\workspace\windows-TGrou---2a8f9c29\bin/packages/Release/specs/runtime.win-x86.Microsoft.NETCore.Targets.nuspec
  Microsoft.NETCore.Platforms -> D:\j\workspace\windows-TGrou---2a8f9c29\bin/packages/Release/specs/runtime.win-x86.Microsoft.NETCore.Platforms.nuspec
  Microsoft.Private.PackageBaseline -> D:\j\workspace\windows-TGrou---2a8f9c29\bin/packages/Release/specs/runtime.win.Microsoft.Private.PackageBaseline.nuspec
  Microsoft.Private.PackageBaseline -> D:\j\workspace\windows-TGrou---2a8f9c29\bin/packages/Release/specs/runtime.win-x86.Microsoft.Private.PackageBaseline.nuspec
  Microsoft.Private.PackageBaseline -> D:\j\workspace\windows-TGrou---2a8f9c29\bin/packages/Release/specs/runtime.win.Microsoft.Private.PackageBaseline.nuspec
  Microsoft.Private.PackageBaseline -> D:\j\workspace\windows-TGrou---2a8f9c29\bin/packages/Release/specs/runtime.win-x86.Microsoft.Private.PackageBaseline.nuspec
  Microsoft.Private.PackageBaseline -> D:\j\workspace\windows-TGrou---2a8f9c29\bin/packages/Release/specs/runtime.win.Microsoft.Private.PackageBaseline.nuspec
  Microsoft.Private.PackageBaseline -> D:\j\workspace\windows-TGrou---2a8f9c29\bin/packages/Release/specs/runtime.win-x86.Microsoft.Private.PackageBaseline.nuspec
  NETStandard.Library.NETFramework -> D:\j\workspace\windows-TGrou---2a8f9c29\bin/packages/Release/specs/NETStandard.Library.NETFramework.nuspec
```
We shouldn't be building those packages multiple times, nor with RIDs set.  Likely something that needs to be addressed in targets in the pkg folder.  Haven't looked deeper yet.

tmds commented:
Probably it gets triggered because PackageRID is set.

ericstj commented on pkg/dir.props:
```diff
@@ -2,11 +2,6 @@
-    <_runtimeOSFamily Condition="'$(_runtimeOSVersionIndex)' != '-1'">$(RuntimeOS.SubString(0, $(_runtimeOSVersionIndex)))</_runtimeOSFamily>
-  </PropertyGroup>
-
   <!-- Packages opt-in to automatic RID-specific builds by placing a *.RID.props next to their project
```
This is what's causing your build failures.  You've redefined this contract.  Please update the comment and decide what the new convention will be.  Perhaps its the existence of the props file /cc @joperezr @chcosta

tmds replied:
I don't understand what mean.
These properties are used to calculate PackageRID. This all moved up one level.

My guess (which may be off) is that moving PackageRID to the top is causing the NETFX build failures.
Perhaps it has to do with this in `pkg/dir.pros`:
```xml
  <ItemGroup>
    <BuildRID Include="@(OfficialBuildRID)" Exclude="$(PackageRID)"/>
    <BuildRID Include="$(PackageRID)">
      <Platform Condition="'$(ArchGroup)' == 'x64'">amd64</Platform>
      <Platform Condition="'$(ArchGroup)' != 'x64'">$(ArchGroup)</Platform>
    </BuildRID>
  </ItemGroup>
```

ericstj replied:
Please read the comment.  Prior to your change PackageRID was treated optional and meant to signify opt-in to RID-specific packages.  Now you're setting PackageRID for everything so it can no longer be opt-in. 

I suggest you change this:
```xml
  <!-- Packages opt-in to automatic RID-specific builds by placing a *.RID.props next to their project
       that defines the following:
       PackageRID property: the RID which we should be building for
       OfficialBuildRID item: all RIDs targeted by the package -->
  <Import Project="$(MSBuildProjectDirectory)\*.rids.props" />

  <!-- create the "BuildRID" item which is the set of all supported RIDs, with metadata.
       We'll add a RID for the current platform even if it isn't in the officially supported set -->
  <ItemGroup>
    <BuildRID Include="@(OfficialBuildRID)" Exclude="$(PackageRID)"/>
    <BuildRID Include="$(PackageRID)">
      <Platform Condition="'$(ArchGroup)' == 'x64'">amd64</Platform>
      <Platform Condition="'$(ArchGroup)' != 'x64'">$(ArchGroup)</Platform>
    </BuildRID>
  </ItemGroup>
```

To this:
```xml
  <!-- Packages opt-in to automatic RID-specific builds by placing a *.RID.props next to their project
       that defines the OfficialBuildRID item: all RIDs targeted by the package -->
  <Import Project="$(MSBuildProjectDirectory)\*.rids.props" />

  <!-- create the "BuildRID" item which is the set of all supported RIDs, with metadata.
       We'll add a RID for the current platform even if it isn't in the officially supported set -->
  <ItemGroup Condition="'@(OfficialBuildRID)' != ''">
    <BuildRID Include="@(OfficialBuildRID)" Exclude="$(PackageRID)"/>
    <BuildRID Include="$(PackageRID)">
      <Platform Condition="'$(ArchGroup)' == 'x64'">amd64</Platform>
      <Platform Condition="'$(ArchGroup)' != 'x64'">$(ArchGroup)</Platform>
    </BuildRID>
  </ItemGroup>
```

tmds commented:
🎉 CI builds
@ericstj @joperezr thank you for reviewing (and patience). Can you have another look at the PR?

ericstj commented on dir.props:
```diff
@@ -117,6 +117,27 @@
+    <_portableRID Condition="'$(_runtimeOSFamily)' == 'win'">win</_portableRID>
+    <_portableRID Condition="'$(_runtimeOSFamily)' == 'osx'">osx</_portableRID>
+    <!-- support cross-targeting by choosing a RID to restore when running on a different machine that what we're build for -->
+    <_portableRID Condition="'$(OSGroup)' == 'Unix' AND '$(_runtimeOSFamily)' != 'osx'">linux</_portableRID>
```
These look a little different than https://github.com/dotnet/corefx/pull/21032/files#diff-2c6964675d4545109b9addd11497c910L6.  Also the comment above is a bit misplaced now.  The property you are setting is not just used for restore, but also for the build of packages.  Given that this logic is really specific to the runtime.depproj it might make sense to keep it there to reduce risk / number of inputs that determine PackageRID.

tmds replied:
I'm not sure how cross-targeting works and if/how it relates to portable builds. Can you give an example of how to cross-target on a Linux machine? I can move it back to runtime.depproj but then it will no longer affect the package name.

ericstj commented on dir.props:
```diff
@@ -117,6 +117,27 @@
+
+    <_packageRID/>
+    <_packageRID Condition="'$(PortableBuild)' == 'true'">$(_portableRID)-$(ArchGroup)</_packageRID>
+    <_packageRID Condition="'$(TargetGroup)' == 'uap'">win10-$(ArchGroup)</_packageRID>
```
Similarly, this logic is specific to the UAP meta-package projects so it may make more sense just to keep it in those projects. /cc @joperezr

tmds replied:
`runtime.depproj` relies on this being set.

tmds commented:
@ericstj `bin\packages\Debug\specs`, there is no such folder/file when I do a build.sh.

tmds commented:
I compared all paths under bin before and after this PR for `build.sh`, there are no significant differences.
When I try the `allConfigurations` build, the _before_ fails with this message:
```
/home/tmds/repos/corefx/dir.props(209,5): error MSB4184: The expression ""uapaot".TrimEnd(aot)" cannot be evaluated. String must be exactly one character long.
```
this is referring to this line:
```
<RefPath Condition="$(TargetGroup.EndsWith('aot'))">$(RefRootPath)$(TargetGroup.TrimEnd('aot'))/</RefPath>
```

ericstj commented:
> there is no such folder/file when I do a build.sh.

It's important that you examine the result of the package build.  From your previous example it looks like you were doing a release build, in which case the path would be `bin/packages/Release/specs`.  Please take care that you are diffing the built nuspecs, use your best judgement to correct the path to whatever is appropriate to your machine.

The error you are getting in the allConfigurations build is something I've seen before on different MSBuild versions.  In some cases they'll pick the char[] overload but not always.  Regardless, that is bad usage of trim end.  Can you replace it with `$(TargetGroup.Remove($(TargetGroup.LastIndexOf('aot'))))` and see if that unblocks you? /cc @joperezr

joperezr commented:
@ericstj is the reason why we haven't noticed this before is because we never tried build.sh -allConfigurations before?

ericstj commented:
Yeah, none of the CIs run allConfigurations on unix nor do the official builds.

joperezr commented:
I see, I'm fine with the change for removing 'aot' as a workaround, initially when I added that we only really meant to use it on Windows, since we were not expecting to build uapaot on non-Windows boxes

tmds commented:
@ericstj I rebased against master and compared the spec files for build.sh. The diff looks good.
I am unable to build allConfigurations on my Linux box, even not before this PR with the aot change. Perhaps this doesn't work on Linux?

You gave the suggestion to move the uap logic and the cross-target logic at the package and runtime.depproj respectively. They are at the toplevel now because they are used both for restoring and to name the package. Does that make sense?

weshaggard commented on dir.props:
```diff
@@ -124,6 +124,27 @@
+
+    <_runtimeOS>$(RuntimeOS)</_runtimeOS>
+    <_runtimeOS Condition="'$(_runtimeOS)' == 'tizen.4.0.0'">ubuntu.14.04</_runtimeOS>
+    <RuntimeRID>$(_runtimeOS)-x64</RuntimeRID>
```
Why are you hardcoding x64 here?

weshaggard replied:
Looks like this is more limited in scope then I had thought it looks like is only for our tools. We should switch to using the portable RID here and think about whether or not we want to hard code it to x64. If we do perhaps we should change this to be ToolRuntimeRID to indicate it is only about tools.

tmds replied:
> We should switch to using the portable RID here

Yes, when doing a portable build. When a maintainer does a distro specific build, he may want to use his own runtime to achieve that build.

> think about whether or not we want to hard code it to x64

This value should be provided by BuildTools then (like RuntimeOS).

> perhaps we should change this to be ToolRuntimeRID

I also think RuntimeRID doesn't convey the meaning well. To properly rename it, the change should also be made to the Buildtools and corefx config, since those generate/accept RuntimeOS.
To suggest another alternative: HostRID. (Buildtools could provide HostOS and HostArch.)

weshaggard replied:
>Yes, when doing a portable build. When a maintainer does a distro specific build, he may want to use his own runtime to achieve that build.

That is fine we would still guard it via the PortableBuild property just like we do in the other places. If it isn't portable then we will use the specific RID.

I still think the name Host is still a overloaded. I agree these are properties that should be set as part of tool initialization but it isn't that straight forward as an individual repo could chose to use different architectures for the CLI they use to build. I still like calling this ToolRuntimeRID as it more closely says what it is for.

weshaggard commented:
>I am unable to build allConfigurations on my Linux box, even not before this PR with the aot change. Perhaps this doesn't work on Linux?

We should probably file an issue to fix this s we want the package build to be able to happen on non-windows as well.

weshaggard commented on external/ilasm/ilasm.depproj:
```diff
@@ -2,10 +2,7 @@
-    <NugetRuntimeIdentifier>$(RuntimeOS)-x64</NugetRuntimeIdentifier>
-    <!-- Tizen does not provide the x64 runtime so replaces it with the ubuntu.14.04-x64 -->
-    <NugetRuntimeIdentifier Condition="'$(RuntimeOS)' == 'tizen.4.0.0'">ubuntu.14.04-x64</NugetRuntimeIdentifier>
+    <NugetRuntimeIdentifier>$(RuntimeRID)</NugetRuntimeIdentifier>
```
@mellinoe it would appear that we used to depend on this being x64 all the time we need to ensure that is still the case after your changes.

mellinoe replied:
Based on the change above (with your comment), it looks like it will still be x64. Technically we just need to restore an ILAsm that is compatible with the build environment, so an x86 version would/should work for Windows builds. But it's probably best and safest to keep it using the x64 version.

weshaggard commented on config.json:
```diff
@@ -31,11 +31,17 @@
+      "description": "The RID of the target package.",
+      "valueType": "property",
+      "values": ["win7-x64", "ubuntu.14.04-x64"],
+      "defaultValue": "${RuntimeOS}-${ArchGroup}"
```
I suspect this default value isn't doing what you think it does. We don't support a generalized format string here it only supports a few custom strings.

tmds replied:
Ah. I assumed this isn't used. The value I put here is the one that gets defaulted by the project. Should I clear it out?

weshaggard replied:
I would just remove it.

weshaggard commented on config.json:
```diff
@@ -31,11 +31,17 @@
       "defaultValue": ""
     },
+    "PackageRID": {
+      "description": "The RID of the target package.",
```
What is the target package in this case? I assume you are implying that it is the RID of what you are building, however I see you a also using it in restoring dependencies.

tmds replied:
Yes, this is the target RID you are building for.

tmds commented on external/runtime/runtime.depproj:
```diff
@@ -2,11 +2,7 @@
-    <NugetRuntimeIdentifier Condition="'$(OSGroup)' == 'Unix' AND '$(RunningOnUnix)' != 'true'">linux-x64</NugetRuntimeIdentifier>
-    <NugetRuntimeIdentifier Condition="'$(TargetGroup)' == 'uap'">win10-$(ArchGroup)</NugetRuntimeIdentifier>
-    <NugetRuntimeIdentifier Condition="'$(TargetGroup)' == 'uapaot'">win10-$(ArchGroup)-aot</NugetRuntimeIdentifier>
+    <NugetRuntimeIdentifier>$(PackageRID)</NugetRuntimeIdentifier>
```
Doesn't this runtime get included in the built package.
If I am building for 'banana' linux, I want to get&use the 'banana' runtime?

weshaggard replied:
No the runtime doesn't get used for anything beyond running tests in this repo. It doesn't get included in any packages that we produce in corefx.

tmds replied:
I see. I thought Microsoft.NETCore.App is built by corefx? This package contains the runtime?

weshaggard replied:
That package is assembled in the core-setup repo, which consumes the runtime packages and the private corefx package (Microsoft.Private.CoreFx.NETCoreApp) but there are no packages that build in corefx that contain the runtime.

tmds replied:
runtime.depproj pre-PR looks like it doesn't match the host but the target.
e.g.
```
<NugetRuntimeIdentifier Condition="'$(OSGroup)' == 'Windows_NT' AND '$(RunningOnUnix)' == 'true'">win-x64</NugetRuntimeIdentifier>
```

weshaggard replied:
Yes that is a specific scenario where someone is trying to build for Windows while running on Unix. 

I guess I did overstate when I mentioned that this is only used for testing. There are some scenarios were we use the assets for building the product in the case of a ReferenceFromRuntime reference (https://github.com/dotnet/corefx/blob/master/referenceFromRuntime.targets#L3) for projects that directly compile against System.Private.CoreLib. Unfortunately there are some slightly different APIs across the different OS's. 

So the most common scenario is folks that want to build and run on the device the are on but in some cases folks will want to build for another runtime but not run locally but instead send them to another machine to run. So that is more of a target/PackageRID as opposed to Host/ToolRuntimeRID. 

So with that said I think using PackageRID here is appropriate.

tmds commented:
@dotnet-bot test Windows x64 Debug Build please

weshaggard commented:
@tmds I would still like you to rename HostRID to ToolRuntimeRID to make the name a little more clear.

weshaggard commented:
test UWP CoreCLR x64 Debug Build

weshaggard commented:
@tmds Thanks a lot for all the work and iteration on the classifying the different type of RID's we have in our system. We truly appreciate you working through this with us and you are now one of experts with in comes to RID usage in corefx and .NET Core. Thanks for the contribution.
